{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ML Hands-on Challenge - Data preparation and exploration\n",
    "\n",
    "This notebook contains the data preparation and exploration for the ML Hands-on Challenge. The task will be to train a protein structure classification model using the CATH database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading things"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import glob\n",
    "import os\n",
    "import Bio.PDB.PDBParser\n",
    "import py3Dmol\n",
    "import warnings\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", message=\"Used element '.' for Atom\")\n",
    "\n",
    "# Get parent directory and add src/ to pythonpath\n",
    "parent_dir = os.path.dirname(os.getcwd())\n",
    "sys.path.append(parent_dir)\n",
    "\n",
    "from src.config import CONSTANTS\n",
    "from src.pdb import *\n",
    "\n",
    "DATA_HOME = os.path.join(parent_dir, CONSTANTS.DATA_HOME)\n",
    "architecture_names = CONSTANTS.ARCHITECTURE_NAMES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the `csv` dataframe with information about sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6273 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01      3            90      1170           50   \n",
       "1     3be3A00      2            30        30          320   \n",
       "2     3zq4C03      3            10        20          580   \n",
       "3     1peqA03      1            10      1650           20   \n",
       "4     1bdoA00      2            40        50          100   \n",
       "...       ...    ...           ...       ...          ...   \n",
       "6268  2yyiA02      2            40       110           10   \n",
       "6269  4mo0A00      3            30       780           10   \n",
       "6270  1vq8X00      3            10       440           10   \n",
       "6271  1ze3D00      3            10        20          410   \n",
       "6272  3ejvA00      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "0                        2.60   2w3s   \n",
       "1                        2.04   3be3   \n",
       "2                        3.00   3zq4   \n",
       "3                        2.80   1peq   \n",
       "4                        1.80   1bdo   \n",
       "...                       ...    ...   \n",
       "6268                     1.66   2yyi   \n",
       "6269                     2.10   4mo0   \n",
       "6270                     2.20   1vq8   \n",
       "6271                     1.84   1ze3   \n",
       "6272                     1.40   3ejv   \n",
       "\n",
       "                                              sequences  cath_indices  \n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...    [(2, 124)]  \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...     [(6, 81)]  \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...  [(449, 555)]  \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...  [(294, 346)]  \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   [(77, 156)]  \n",
       "...                                                 ...           ...  \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...  [(139, 266)]  \n",
       "6269  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   [(24, 102)]  \n",
       "6270  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...     [(7, 88)]  \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...    [(1, 125)]  \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...    [(2, 160)]  \n",
       "\n",
       "[6273 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 6273 entries, 0 to 6272\n",
      "Data columns (total 9 columns):\n",
      " #   Column                   Non-Null Count  Dtype  \n",
      "---  ------                   --------------  -----  \n",
      " 0   cath_id                  6273 non-null   object \n",
      " 1   class                    6273 non-null   int64  \n",
      " 2   architecture             6273 non-null   int64  \n",
      " 3   topology                 6273 non-null   int64  \n",
      " 4   superfamily              6273 non-null   int64  \n",
      " 5   resolution_in_angstroms  6273 non-null   float64\n",
      " 6   pdb_id                   6273 non-null   object \n",
      " 7   sequences                6268 non-null   object \n",
      " 8   cath_indices             6268 non-null   object \n",
      "dtypes: float64(1), int64(4), object(4)\n",
      "memory usage: 490.1+ KB\n"
     ]
    }
   ],
   "source": [
    "# Open the training data sequences and structure\n",
    "data = pd.read_csv(f\"{DATA_HOME}/cath_w_seqs_share.csv\", index_col=0)\n",
    "display(data)\n",
    "\n",
    "# Get the schema of the data\n",
    "data.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Converting `cath_indices` to actual CATH indices using `eval`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2h/yv02dgq91vd9r1zfwcgl5j3r0000gp/T/ipykernel_83746/681847689.py:6: UserWarning: Error occurred while evaluating nan: eval() arg 1 must be a string, bytes or code object\n",
      "  warnings.warn(f\"Error occurred while evaluating {x}: {e}\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6273 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01      3            90      1170           50   \n",
       "1     3be3A00      2            30        30          320   \n",
       "2     3zq4C03      3            10        20          580   \n",
       "3     1peqA03      1            10      1650           20   \n",
       "4     1bdoA00      2            40        50          100   \n",
       "...       ...    ...           ...       ...          ...   \n",
       "6268  2yyiA02      2            40       110           10   \n",
       "6269  4mo0A00      3            30       780           10   \n",
       "6270  1vq8X00      3            10       440           10   \n",
       "6271  1ze3D00      3            10        20          410   \n",
       "6272  3ejvA00      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "0                        2.60   2w3s   \n",
       "1                        2.04   3be3   \n",
       "2                        3.00   3zq4   \n",
       "3                        2.80   1peq   \n",
       "4                        1.80   1bdo   \n",
       "...                       ...    ...   \n",
       "6268                     1.66   2yyi   \n",
       "6269                     2.10   4mo0   \n",
       "6270                     2.20   1vq8   \n",
       "6271                     1.84   1ze3   \n",
       "6272                     1.40   3ejv   \n",
       "\n",
       "                                              sequences  cath_indices  \\\n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...    [(2, 124)]   \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...     [(6, 81)]   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...  [(449, 555)]   \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...  [(294, 346)]   \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   [(77, 156)]   \n",
       "...                                                 ...           ...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...  [(139, 266)]   \n",
       "6269  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   [(24, 102)]   \n",
       "6270  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...     [(7, 88)]   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...    [(1, 125)]   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...    [(2, 160)]   \n",
       "\n",
       "     cath_indices_evaled  \n",
       "0             [(2, 124)]  \n",
       "1              [(6, 81)]  \n",
       "2           [(449, 555)]  \n",
       "3           [(294, 346)]  \n",
       "4            [(77, 156)]  \n",
       "...                  ...  \n",
       "6268        [(139, 266)]  \n",
       "6269         [(24, 102)]  \n",
       "6270           [(7, 88)]  \n",
       "6271          [(1, 125)]  \n",
       "6272          [(2, 160)]  \n",
       "\n",
       "[6273 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply eval to `cath_indices` with raising warning if catching error\n",
    "def apply_eval_with_warning(x):\n",
    "    try:\n",
    "        return eval(x)\n",
    "    except Exception as e:\n",
    "        warnings.warn(f\"Error occurred while evaluating {x}: {e}\")\n",
    "        return [(np.nan, np.nan)]\n",
    "\n",
    "\n",
    "data[\"cath_indices_evaled\"] = data[\"cath_indices\"].apply(apply_eval_with_warning)\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the information from sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Asserting that `cath_id` in the `data` corresponds to files in `pdb_share` folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List how many files are in {DATA}/data/pdb_share\n",
    "pdb_share_dir = os.path.join(DATA_HOME, \"pdb_share\")\n",
    "file_paths = glob.glob(os.path.join(pdb_share_dir, \"*\"))\n",
    "\n",
    "# Get a list of file names\n",
    "file_names = [os.path.basename(file_path) for file_path in file_paths]\n",
    "\n",
    "# Assert that set of `cath_id` from `data` is equal to the set of `file_names`\n",
    "assert set(data[\"cath_id\"]) == set(file_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the structures from pdb files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'SE' for Atom (name=SE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:273: PDBConstructionWarning: WARNING: disordered atom found with blank altloc before line 817.\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'SE' for Atom (name=SE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:273: PDBConstructionWarning: WARNING: disordered atom found with blank altloc before line 817.\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6273 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01      3            90      1170           50   \n",
       "1     3be3A00      2            30        30          320   \n",
       "2     3zq4C03      3            10        20          580   \n",
       "3     1peqA03      1            10      1650           20   \n",
       "4     1bdoA00      2            40        50          100   \n",
       "...       ...    ...           ...       ...          ...   \n",
       "6268  2yyiA02      2            40       110           10   \n",
       "6269  4mo0A00      3            30       780           10   \n",
       "6270  1vq8X00      3            10       440           10   \n",
       "6271  1ze3D00      3            10        20          410   \n",
       "6272  3ejvA00      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "0                        2.60   2w3s   \n",
       "1                        2.04   3be3   \n",
       "2                        3.00   3zq4   \n",
       "3                        2.80   1peq   \n",
       "4                        1.80   1bdo   \n",
       "...                       ...    ...   \n",
       "6268                     1.66   2yyi   \n",
       "6269                     2.10   4mo0   \n",
       "6270                     2.20   1vq8   \n",
       "6271                     1.84   1ze3   \n",
       "6272                     1.40   3ejv   \n",
       "\n",
       "                                              sequences  cath_indices  \\\n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...    [(2, 124)]   \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...     [(6, 81)]   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...  [(449, 555)]   \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...  [(294, 346)]   \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   [(77, 156)]   \n",
       "...                                                 ...           ...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...  [(139, 266)]   \n",
       "6269  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   [(24, 102)]   \n",
       "6270  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...     [(7, 88)]   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...    [(1, 125)]   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...    [(2, 160)]   \n",
       "\n",
       "     cath_indices_evaled                                      pdb_sequences  \\\n",
       "0             [(2, 124)]  SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1              [(6, 81)]  QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2           [(449, 555)]  DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3           [(294, 346)]  DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4            [(77, 156)]  EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "...                  ...                                                ...   \n",
       "6268        [(139, 266)]  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6269         [(24, 102)]  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   \n",
       "6270           [(7, 88)]  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...   \n",
       "6271          [(1, 125)]  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272          [(2, 160)]  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                       pdb_cath_indices  \n",
       "0                                            [(2, 124)]  \n",
       "1                                   [(6, 49), (51, 81)]  \n",
       "2                              [(449, 492), (501, 555)]  \n",
       "3                                          [(294, 346)]  \n",
       "4                                           [(77, 156)]  \n",
       "...                                                 ...  \n",
       "6268                           [(139, 196), (199, 266)]  \n",
       "6269                                        [(24, 102)]  \n",
       "6270                                          [(7, 88)]  \n",
       "6271                                [(1, 9), (19, 125)]  \n",
       "6272  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...  \n",
       "\n",
       "[6273 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = data.assign(\n",
    "    **{\n",
    "        \"pdb_sequences\": data[\"cath_id\"].apply(\n",
    "            lambda x: get_sequence_from_pdb_file(f\"{DATA_HOME}/pdb_share/{x}\")\n",
    "        ),\n",
    "        \"pdb_cath_indices\": data[\"cath_id\"].apply(\n",
    "            lambda x: get_res_list_from_pdb_file(f\"{DATA_HOME}/pdb_share/{x}\")\n",
    "        ),\n",
    "    }\n",
    ")\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Asserting that `sequences` and `pdb_sequences` are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>XXXXXXXXXXXXXXXXXXXXXPHLSEQLCFFVQAREIADFYEKYAL...</td>\n",
       "      <td>[(101, 121), (123, 136), (138, 145), (147, 172...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>2xskA00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2420</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2xsk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>SSQITFNTTQQGDYTIIPEVTLTQSXLXRVQILSLREGSSGQSQTK...</td>\n",
       "      <td>[(3, 15), (17, 97)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4668</th>\n",
       "      <td>4dgwC00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2690</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4dgw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXGSVGAIQVNY...</td>\n",
       "      <td>[(101, 136), (149, 152), (154, 225), (227, 253)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>1nthA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>460</td>\n",
       "      <td>1.55</td>\n",
       "      <td>1nth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>TFRKSFDCYDFYDRAKVGEKCTQDDWDLMKIPMKAMELKQKYGLDF...</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>3pieA02</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>1370</td>\n",
       "      <td>250</td>\n",
       "      <td>2.90</td>\n",
       "      <td>3pie</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>GKKLLMKQQKKLIGAVKPWLLKTVQRKVTSDADFEIFPLEDKELVR...</td>\n",
       "      <td>[(379, 408), (412, 462), (469, 487)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "7     2gnxA01      1            10      3450           30   \n",
       "2827  2xskA00      2            60        40         2420   \n",
       "4668  4dgwC00      2            60        40         2690   \n",
       "5776  1nthA00      3            20        20          460   \n",
       "6010  3pieA02      3            30      1370          250   \n",
       "\n",
       "      resolution_in_angstroms pdb_id sequences              cath_indices  \\\n",
       "7                        2.45   2gnx       NaN  [(101, 295), (431, 440)]   \n",
       "2827                     1.70   2xsk       NaN                 [(3, 97)]   \n",
       "4668                     3.11   4dgw       NaN              [(101, 253)]   \n",
       "5776                     1.55   1nth       NaN                [(2, 458)]   \n",
       "6010                     2.90   3pie       NaN              [(379, 487)]   \n",
       "\n",
       "           cath_indices_evaled  \\\n",
       "7     [(101, 295), (431, 440)]   \n",
       "2827                 [(3, 97)]   \n",
       "4668              [(101, 253)]   \n",
       "5776                [(2, 458)]   \n",
       "6010              [(379, 487)]   \n",
       "\n",
       "                                          pdb_sequences  \\\n",
       "7     XXXXXXXXXXXXXXXXXXXXXPHLSEQLCFFVQAREIADFYEKYAL...   \n",
       "2827  SSQITFNTTQQGDYTIIPEVTLTQSXLXRVQILSLREGSSGQSQTK...   \n",
       "4668  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXGSVGAIQVNY...   \n",
       "5776  TFRKSFDCYDFYDRAKVGEKCTQDDWDLMKIPMKAMELKQKYGLDF...   \n",
       "6010  GKKLLMKQQKKLIGAVKPWLLKTVQRKVTSDADFEIFPLEDKELVR...   \n",
       "\n",
       "                                       pdb_cath_indices  \n",
       "7     [(101, 121), (123, 136), (138, 145), (147, 172...  \n",
       "2827                                [(3, 15), (17, 97)]  \n",
       "4668   [(101, 136), (149, 152), (154, 225), (227, 253)]  \n",
       "5776                                         [(2, 458)]  \n",
       "6010               [(379, 408), (412, 462), (469, 487)]  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch_data = data[data[\"pdb_sequences\"] != data[\"sequences\"]]\n",
    "mismatch_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that there are some mismatch sequences when comparing local pdb files and `sequences` from the `csv` file. Apparently, it is due to unusual / non-standard amino acids. We will remove these sequences from the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dropping the mismatched data\n",
    "seq_filtered_data = data.drop(mismatch_data.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Asserting that `cath_indices` and `pdb_cath_indices` are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1aqcA00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>2.30</td>\n",
       "      <td>1aqc</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...</td>\n",
       "      <td>[(324, 488)]</td>\n",
       "      <td>[(324, 488)]</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...</td>\n",
       "      <td>[(324, 353), (356, 365), (367, 370), (386, 407...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3i9v600</td>\n",
       "      <td>3</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>12280</td>\n",
       "      <td>3.10</td>\n",
       "      <td>3i9v</td>\n",
       "      <td>EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...</td>\n",
       "      <td>[(15, 175)]</td>\n",
       "      <td>[(15, 175)]</td>\n",
       "      <td>EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...</td>\n",
       "      <td>[(15, 57), (74, 175)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3rylA01</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>120</td>\n",
       "      <td>1210</td>\n",
       "      <td>3.10</td>\n",
       "      <td>3ryl</td>\n",
       "      <td>GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...</td>\n",
       "      <td>[(244, 283), (385, 481)]</td>\n",
       "      <td>[(244, 283), (385, 481)]</td>\n",
       "      <td>GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...</td>\n",
       "      <td>[(244, 245), (247, 283), (385, 411), (413, 448...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6266</th>\n",
       "      <td>1n99A02</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>1.94</td>\n",
       "      <td>1n99</td>\n",
       "      <td>RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...</td>\n",
       "      <td>[(197, 269)]</td>\n",
       "      <td>[(197, 269)]</td>\n",
       "      <td>RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...</td>\n",
       "      <td>[(197, 200), (202, 269)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6267</th>\n",
       "      <td>1jceA03</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>640</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>1jce</td>\n",
       "      <td>AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...</td>\n",
       "      <td>[(178, 253)]</td>\n",
       "      <td>[(178, 253)]</td>\n",
       "      <td>AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...</td>\n",
       "      <td>[(178, 181), (183, 253)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2707 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "1     3be3A00      2            30        30          320   \n",
       "2     3zq4C03      3            10        20          580   \n",
       "6     1aqcA00      2            30        29           30   \n",
       "12    3i9v600      3            40        50        12280   \n",
       "13    3rylA01      1            20       120         1210   \n",
       "...       ...    ...           ...       ...          ...   \n",
       "6266  1n99A02      2            30        42           10   \n",
       "6267  1jceA03      3            90       640           10   \n",
       "6268  2yyiA02      2            40       110           10   \n",
       "6271  1ze3D00      3            10        20          410   \n",
       "6272  3ejvA00      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "1                        2.04   3be3   \n",
       "2                        3.00   3zq4   \n",
       "6                        2.30   1aqc   \n",
       "12                       3.10   3i9v   \n",
       "13                       3.10   3ryl   \n",
       "...                       ...    ...   \n",
       "6266                     1.94   1n99   \n",
       "6267                     2.10   1jce   \n",
       "6268                     1.66   2yyi   \n",
       "6271                     1.84   1ze3   \n",
       "6272                     1.40   3ejv   \n",
       "\n",
       "                                              sequences  \\\n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "6     EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...   \n",
       "12    EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...   \n",
       "13    GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...   \n",
       "...                                                 ...   \n",
       "6266  RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...   \n",
       "6267  AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                  cath_indices       cath_indices_evaled  \\\n",
       "1                    [(6, 81)]                 [(6, 81)]   \n",
       "2                 [(449, 555)]              [(449, 555)]   \n",
       "6                 [(324, 488)]              [(324, 488)]   \n",
       "12                 [(15, 175)]               [(15, 175)]   \n",
       "13    [(244, 283), (385, 481)]  [(244, 283), (385, 481)]   \n",
       "...                        ...                       ...   \n",
       "6266              [(197, 269)]              [(197, 269)]   \n",
       "6267              [(178, 253)]              [(178, 253)]   \n",
       "6268              [(139, 266)]              [(139, 266)]   \n",
       "6271                [(1, 125)]                [(1, 125)]   \n",
       "6272                [(2, 160)]                [(2, 160)]   \n",
       "\n",
       "                                          pdb_sequences  \\\n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "6     EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...   \n",
       "12    EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...   \n",
       "13    GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...   \n",
       "...                                                 ...   \n",
       "6266  RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...   \n",
       "6267  AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                       pdb_cath_indices  \n",
       "1                                   [(6, 49), (51, 81)]  \n",
       "2                              [(449, 492), (501, 555)]  \n",
       "6     [(324, 353), (356, 365), (367, 370), (386, 407...  \n",
       "12                                [(15, 57), (74, 175)]  \n",
       "13    [(244, 245), (247, 283), (385, 411), (413, 448...  \n",
       "...                                                 ...  \n",
       "6266                           [(197, 200), (202, 269)]  \n",
       "6267                           [(178, 181), (183, 253)]  \n",
       "6268                           [(139, 196), (199, 266)]  \n",
       "6271                                [(1, 9), (19, 125)]  \n",
       "6272  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...  \n",
       "\n",
       "[2707 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch_rows = seq_filtered_data[\n",
    "    seq_filtered_data[\"cath_indices_evaled\"] != seq_filtered_data[\"pdb_cath_indices\"]\n",
    "]\n",
    "mismatch_rows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that around 2.7k sequences have gaps in the sequences that are depicted in the `pdb_cath_indices` but not in the `cath_indices`. It would be interesting to have a look at the sequence lengths."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculating the length of the sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "      <th>sequences_len</th>\n",
       "      <th>pdb_sequences_len</th>\n",
       "      <th>cath_indices_len</th>\n",
       "      <th>pdb_cath_indices_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>123.0</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>76.0</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>107.0</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>53.0</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>80.0</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>[(139, 266)]</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>128.0</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82.0</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>[(1, 125)]</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "      <td>116</td>\n",
       "      <td>116</td>\n",
       "      <td>125.0</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>[(2, 160)]</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "      <td>152</td>\n",
       "      <td>152</td>\n",
       "      <td>159.0</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6268 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01      3            90      1170           50   \n",
       "1     3be3A00      2            30        30          320   \n",
       "2     3zq4C03      3            10        20          580   \n",
       "3     1peqA03      1            10      1650           20   \n",
       "4     1bdoA00      2            40        50          100   \n",
       "...       ...    ...           ...       ...          ...   \n",
       "6268  2yyiA02      2            40       110           10   \n",
       "6269  4mo0A00      3            30       780           10   \n",
       "6270  1vq8X00      3            10       440           10   \n",
       "6271  1ze3D00      3            10        20          410   \n",
       "6272  3ejvA00      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "0                        2.60   2w3s   \n",
       "1                        2.04   3be3   \n",
       "2                        3.00   3zq4   \n",
       "3                        2.80   1peq   \n",
       "4                        1.80   1bdo   \n",
       "...                       ...    ...   \n",
       "6268                     1.66   2yyi   \n",
       "6269                     2.10   4mo0   \n",
       "6270                     2.20   1vq8   \n",
       "6271                     1.84   1ze3   \n",
       "6272                     1.40   3ejv   \n",
       "\n",
       "                                              sequences  cath_indices  \\\n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...    [(2, 124)]   \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...     [(6, 81)]   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...  [(449, 555)]   \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...  [(294, 346)]   \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   [(77, 156)]   \n",
       "...                                                 ...           ...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...  [(139, 266)]   \n",
       "6269  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   [(24, 102)]   \n",
       "6270  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...     [(7, 88)]   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...    [(1, 125)]   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...    [(2, 160)]   \n",
       "\n",
       "     cath_indices_evaled                                      pdb_sequences  \\\n",
       "0             [(2, 124)]  SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1              [(6, 81)]  QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2           [(449, 555)]  DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3           [(294, 346)]  DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4            [(77, 156)]  EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "...                  ...                                                ...   \n",
       "6268        [(139, 266)]  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6269         [(24, 102)]  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   \n",
       "6270           [(7, 88)]  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...   \n",
       "6271          [(1, 125)]  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272          [(2, 160)]  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                       pdb_cath_indices  sequences_len  \\\n",
       "0                                            [(2, 124)]            123   \n",
       "1                                   [(6, 49), (51, 81)]             75   \n",
       "2                              [(449, 492), (501, 555)]             99   \n",
       "3                                          [(294, 346)]             53   \n",
       "4                                           [(77, 156)]             80   \n",
       "...                                                 ...            ...   \n",
       "6268                           [(139, 196), (199, 266)]            126   \n",
       "6269                                        [(24, 102)]             79   \n",
       "6270                                          [(7, 88)]             82   \n",
       "6271                                [(1, 9), (19, 125)]            116   \n",
       "6272  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...            152   \n",
       "\n",
       "      pdb_sequences_len  cath_indices_len  pdb_cath_indices_len  \n",
       "0                   123             123.0                   123  \n",
       "1                    75              76.0                    75  \n",
       "2                    99             107.0                    99  \n",
       "3                    53              53.0                    53  \n",
       "4                    80              80.0                    80  \n",
       "...                 ...               ...                   ...  \n",
       "6268                126             128.0                   126  \n",
       "6269                 79              79.0                    79  \n",
       "6270                 82              82.0                    82  \n",
       "6271                116             125.0                   116  \n",
       "6272                152             159.0                   152  \n",
       "\n",
       "[6268 rows x 16 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_data = seq_filtered_data.assign(\n",
    "    **{\n",
    "        \"sequences_len\": seq_filtered_data[\"sequences\"].apply(len),\n",
    "        \"pdb_sequences_len\": seq_filtered_data[\"pdb_sequences\"].apply(len),\n",
    "        \"cath_indices_len\": seq_filtered_data[\"cath_indices_evaled\"].apply(\n",
    "            lambda lst: np.sum([end - start + 1 for start, end in lst])\n",
    "        ),\n",
    "        \"pdb_cath_indices_len\": seq_filtered_data[\"pdb_cath_indices\"].apply(\n",
    "            lambda lst: np.sum([end - start + 1 for start, end in lst])\n",
    "        ),\n",
    "    }\n",
    ")\n",
    "filtered_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great! We see how `path_cath_indices_len` almost matches the `sequences_len` and `pdb_sequences_len`. Let's confirm that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "      <th>sequences_len</th>\n",
       "      <th>pdb_sequences_len</th>\n",
       "      <th>cath_indices_len</th>\n",
       "      <th>pdb_cath_indices_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cath_id, class, architecture, topology, superfamily, resolution_in_angstroms, pdb_id, sequences, cath_indices, cath_indices_evaled, pdb_sequences, pdb_cath_indices, sequences_len, pdb_sequences_len, cath_indices_len, pdb_cath_indices_len]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>cath_indices_evaled</th>\n",
       "      <th>pdb_sequences</th>\n",
       "      <th>pdb_cath_indices</th>\n",
       "      <th>sequences_len</th>\n",
       "      <th>pdb_sequences_len</th>\n",
       "      <th>cath_indices_len</th>\n",
       "      <th>pdb_cath_indices_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cath_id, class, architecture, topology, superfamily, resolution_in_angstroms, pdb_id, sequences, cath_indices, cath_indices_evaled, pdb_sequences, pdb_cath_indices, sequences_len, pdb_sequences_len, cath_indices_len, pdb_cath_indices_len]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatched_rows = filtered_data[\n",
    "    filtered_data[\"pdb_cath_indices_len\"] != filtered_data[\"sequences_len\"]\n",
    "]\n",
    "display(mismatched_rows)\n",
    "\n",
    "mismatched_rows = filtered_data[\n",
    "    filtered_data[\"pdb_cath_indices_len\"] != filtered_data[\"pdb_sequences_len\"]\n",
    "]\n",
    "mismatched_rows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two empty sets! Great so it means:\n",
    "- The `cath_indices` contains the global boundaries of the CATH domains or supersets of the local boundaries (i.e. the coordinates of the whole sequence)\n",
    "- The `pdb_cath_indices` contains the local boundaries of the CATH domains (i.e. the coordinates of each piece)\n",
    "\n",
    "Let's rename the columns to make it clearer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>sequence</th>\n",
       "      <th>piece_edges</th>\n",
       "      <th>num_residues</th>\n",
       "      <th>domain_edges</th>\n",
       "      <th>total_num_residues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>123</td>\n",
       "      <td>[2, 124]</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>3be3</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "      <td>75</td>\n",
       "      <td>[6, 81]</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "      <td>99</td>\n",
       "      <td>[449, 555]</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1peq</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>53</td>\n",
       "      <td>[294, 346]</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>80</td>\n",
       "      <td>[77, 156]</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "      <td>126</td>\n",
       "      <td>[139, 266]</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>79</td>\n",
       "      <td>[24, 102]</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>82</td>\n",
       "      <td>[7, 88]</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "      <td>116</td>\n",
       "      <td>[1, 125]</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "      <td>152</td>\n",
       "      <td>[2, 160]</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6268 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id pdb_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01   2w3s      3            90      1170           50   \n",
       "1     3be3A00   3be3      2            30        30          320   \n",
       "2     3zq4C03   3zq4      3            10        20          580   \n",
       "3     1peqA03   1peq      1            10      1650           20   \n",
       "4     1bdoA00   1bdo      2            40        50          100   \n",
       "...       ...    ...    ...           ...       ...          ...   \n",
       "6268  2yyiA02   2yyi      2            40       110           10   \n",
       "6269  4mo0A00   4mo0      3            30       780           10   \n",
       "6270  1vq8X00   1vq8      3            10       440           10   \n",
       "6271  1ze3D00   1ze3      3            10        20          410   \n",
       "6272  3ejvA00   3ejv      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms  \\\n",
       "0                        2.60   \n",
       "1                        2.04   \n",
       "2                        3.00   \n",
       "3                        2.80   \n",
       "4                        1.80   \n",
       "...                       ...   \n",
       "6268                     1.66   \n",
       "6269                     2.10   \n",
       "6270                     2.20   \n",
       "6271                     1.84   \n",
       "6272                     1.40   \n",
       "\n",
       "                                               sequence  \\\n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "...                                                 ...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6269  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   \n",
       "6270  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                            piece_edges  num_residues  \\\n",
       "0                                            [(2, 124)]           123   \n",
       "1                                   [(6, 49), (51, 81)]            75   \n",
       "2                              [(449, 492), (501, 555)]            99   \n",
       "3                                          [(294, 346)]            53   \n",
       "4                                           [(77, 156)]            80   \n",
       "...                                                 ...           ...   \n",
       "6268                           [(139, 196), (199, 266)]           126   \n",
       "6269                                        [(24, 102)]            79   \n",
       "6270                                          [(7, 88)]            82   \n",
       "6271                                [(1, 9), (19, 125)]           116   \n",
       "6272  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...           152   \n",
       "\n",
       "     domain_edges  total_num_residues  \n",
       "0        [2, 124]                 123  \n",
       "1         [6, 81]                  76  \n",
       "2      [449, 555]                 107  \n",
       "3      [294, 346]                  53  \n",
       "4       [77, 156]                  80  \n",
       "...           ...                 ...  \n",
       "6268   [139, 266]                 128  \n",
       "6269    [24, 102]                  79  \n",
       "6270      [7, 88]                  82  \n",
       "6271     [1, 125]                 125  \n",
       "6272     [2, 160]                 159  \n",
       "\n",
       "[6268 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_to_keep = [\n",
    "    \"cath_id\",\n",
    "    \"pdb_id\",\n",
    "    \"class\",\n",
    "    \"architecture\",\n",
    "    \"topology\",\n",
    "    \"superfamily\",\n",
    "    \"resolution_in_angstroms\",\n",
    "    \"sequence\",\n",
    "    \"piece_edges\",\n",
    "    \"num_residues\",\n",
    "    \"domain_edges\",\n",
    "    \"total_num_residues\",\n",
    "]\n",
    "\n",
    "proc_data = filtered_data.rename(\n",
    "    columns={\n",
    "        \"pdb_cath_indices\": \"piece_edges\",  # edge indices of each piece\n",
    "        \"sequences_len\": \"num_residues\",  # present/resolved residues in the sequence\n",
    "        \"sequences\": \"sequence\",\n",
    "    }\n",
    ").assign(\n",
    "    **{\n",
    "        \"domain_edges\": lambda df: df[\"piece_edges\"].apply(\n",
    "            lambda lst: [lst[0][0], lst[-1][1]]\n",
    "        ),\n",
    "        \"total_num_residues\": lambda df: df[\"domain_edges\"].apply(\n",
    "            lambda lst: lst[1] - lst[0] + 1\n",
    "        ),\n",
    "    }\n",
    ")[cols_to_keep]\n",
    "\n",
    "proc_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Optional] Recovering missing residues\n",
    "\n",
    "Let us look at the `piece_edges` and understand the prevalence of gaps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "num_gaps\n",
       "0     3037\n",
       "1     1433\n",
       "2      680\n",
       "3      406\n",
       "4      236\n",
       "5      164\n",
       "6      105\n",
       "7       62\n",
       "8       46\n",
       "9       37\n",
       "11      21\n",
       "10      18\n",
       "12       6\n",
       "15       5\n",
       "14       5\n",
       "13       4\n",
       "19       1\n",
       "16       1\n",
       "18       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_data = proc_data.assign(\n",
    "    **{\n",
    "        'num_gaps': proc_data['piece_edges'].apply(lambda x: len(x) - 1),\n",
    "    }\n",
    ")\n",
    "\n",
    "proc_data['num_gaps'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will try to download the protein structures from the PDB database and see if we can recover the missing residues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>sequence</th>\n",
       "      <th>piece_edges</th>\n",
       "      <th>num_residues</th>\n",
       "      <th>domain_edges</th>\n",
       "      <th>total_num_residues</th>\n",
       "      <th>num_gaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>3be3</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "      <td>75</td>\n",
       "      <td>[6, 81]</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "      <td>99</td>\n",
       "      <td>[449, 555]</td>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1aqcA00</td>\n",
       "      <td>1aqc</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>2.30</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...</td>\n",
       "      <td>[(324, 353), (356, 365), (367, 370), (386, 407...</td>\n",
       "      <td>121</td>\n",
       "      <td>[324, 488]</td>\n",
       "      <td>165</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3i9v600</td>\n",
       "      <td>3i9v</td>\n",
       "      <td>3</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>12280</td>\n",
       "      <td>3.10</td>\n",
       "      <td>EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...</td>\n",
       "      <td>[(15, 57), (74, 175)]</td>\n",
       "      <td>145</td>\n",
       "      <td>[15, 175]</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3rylA01</td>\n",
       "      <td>3ryl</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>120</td>\n",
       "      <td>1210</td>\n",
       "      <td>3.10</td>\n",
       "      <td>GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...</td>\n",
       "      <td>[(244, 245), (247, 283), (385, 411), (413, 448...</td>\n",
       "      <td>134</td>\n",
       "      <td>[244, 481]</td>\n",
       "      <td>238</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6266</th>\n",
       "      <td>1n99A02</td>\n",
       "      <td>1n99</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>1.94</td>\n",
       "      <td>RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...</td>\n",
       "      <td>[(197, 200), (202, 269)]</td>\n",
       "      <td>72</td>\n",
       "      <td>[197, 269]</td>\n",
       "      <td>73</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6267</th>\n",
       "      <td>1jceA03</td>\n",
       "      <td>1jce</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>640</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...</td>\n",
       "      <td>[(178, 181), (183, 253)]</td>\n",
       "      <td>75</td>\n",
       "      <td>[178, 253]</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "      <td>126</td>\n",
       "      <td>[139, 266]</td>\n",
       "      <td>128</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "      <td>116</td>\n",
       "      <td>[1, 125]</td>\n",
       "      <td>125</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "      <td>152</td>\n",
       "      <td>[2, 160]</td>\n",
       "      <td>159</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3231 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id pdb_id  class  architecture  topology  superfamily  \\\n",
       "1     3be3A00   3be3      2            30        30          320   \n",
       "2     3zq4C03   3zq4      3            10        20          580   \n",
       "6     1aqcA00   1aqc      2            30        29           30   \n",
       "12    3i9v600   3i9v      3            40        50        12280   \n",
       "13    3rylA01   3ryl      1            20       120         1210   \n",
       "...       ...    ...    ...           ...       ...          ...   \n",
       "6266  1n99A02   1n99      2            30        42           10   \n",
       "6267  1jceA03   1jce      3            90       640           10   \n",
       "6268  2yyiA02   2yyi      2            40       110           10   \n",
       "6271  1ze3D00   1ze3      3            10        20          410   \n",
       "6272  3ejvA00   3ejv      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms  \\\n",
       "1                        2.04   \n",
       "2                        3.00   \n",
       "6                        2.30   \n",
       "12                       3.10   \n",
       "13                       3.10   \n",
       "...                       ...   \n",
       "6266                     1.94   \n",
       "6267                     2.10   \n",
       "6268                     1.66   \n",
       "6271                     1.84   \n",
       "6272                     1.40   \n",
       "\n",
       "                                               sequence  \\\n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "6     EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...   \n",
       "12    EREGILFTTLEKLVAWGRSNSLWPATFGLACCAIEMMASTDARQAD...   \n",
       "13    GHRLLSEDLFKQSPKLSEQELDELANNLADYLFQAADIDISTEEKN...   \n",
       "...                                                 ...   \n",
       "6266  RTITHKDSTGHVGFIFKNGKITSIVKDSSAARNGLLTEHNICEING...   \n",
       "6267  AGDEDEAIVQYVRETYRVAIGERTAERVKIEIGNVFPSKENDELET...   \n",
       "6268  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6272  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                            piece_edges  num_residues  \\\n",
       "1                                   [(6, 49), (51, 81)]            75   \n",
       "2                              [(449, 492), (501, 555)]            99   \n",
       "6     [(324, 353), (356, 365), (367, 370), (386, 407...           121   \n",
       "12                                [(15, 57), (74, 175)]           145   \n",
       "13    [(244, 245), (247, 283), (385, 411), (413, 448...           134   \n",
       "...                                                 ...           ...   \n",
       "6266                           [(197, 200), (202, 269)]            72   \n",
       "6267                           [(178, 181), (183, 253)]            75   \n",
       "6268                           [(139, 196), (199, 266)]           126   \n",
       "6271                                [(1, 9), (19, 125)]           116   \n",
       "6272  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...           152   \n",
       "\n",
       "     domain_edges  total_num_residues  num_gaps  \n",
       "1         [6, 81]                  76         1  \n",
       "2      [449, 555]                 107         1  \n",
       "6      [324, 488]                 165         7  \n",
       "12      [15, 175]                 161         1  \n",
       "13     [244, 481]                 238         4  \n",
       "...           ...                 ...       ...  \n",
       "6266   [197, 269]                  73         1  \n",
       "6267   [178, 253]                  76         1  \n",
       "6268   [139, 266]                 128         1  \n",
       "6271     [1, 125]                 125         1  \n",
       "6272     [2, 160]                 159         6  \n",
       "\n",
       "[3231 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extracting sequences with missing residues\n",
    "miss_residue_pdbs = proc_data[proc_data['num_gaps'] > 0]\n",
    "miss_residue_pdbs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.2k sequences have gaps!\n",
    "\n",
    "<font color='red'>The cell below can run for around ~30 minutes as it downloads all the sequences with gaps!</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "LOAD_PDB = False  # Set to True to download PDB files\n",
    "\n",
    "if LOAD_PDB:\n",
    "    for i, row in tqdm(miss_residue_pdbs.iterrows(), total=miss_residue_pdbs.shape[0]):\n",
    "        download_pdb(row['pdb_id'], f\"{DATA_HOME}/pdb_downloaded\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching the newly downloaded sequences and the original ones\n",
    "\n",
    "Now as we downloaded the new sequences, let's see if we can match them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence_from_pdb_file(pdb_filename, nonstandard_warnings=False):\n",
    "    \"\"\"\n",
    "    Retrieves the protein sequence from a PDB file.\n",
    "\n",
    "    Args:\n",
    "        pdb_filename (str): The path to the PDB file.\n",
    "        nonstandard_warnings (bool, optional): If True, warning messages for nonstandard residues.\n",
    "            Defaults to False.\n",
    "\n",
    "    Returns:\n",
    "        str: The protein sequence extracted from the PDB file.\n",
    "    \"\"\"\n",
    "    pdb_parser = Bio.PDB.PDBParser()\n",
    "    structure = pdb_parser.get_structure(pdb_filename, pdb_filename)\n",
    "    assert len(structure) == 1\n",
    "\n",
    "    seq = []\n",
    "\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            for residue in chain:\n",
    "                if residue.get_id()[0] == \" \":  # This checks if it's a standard residue\n",
    "                    seq.append(protein_letters_3to1[residue.get_resname()])\n",
    "                else:\n",
    "                    if nonstandard_warnings:\n",
    "                        print(\"nonstandard\", residue.get_id())\n",
    "\n",
    "    return \"\".join(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence with gaps info:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>domain_edges</th>\n",
       "      <th>piece_edges</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>[1, 125]</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id pdb_id domain_edges          piece_edges  \\\n",
       "6271  1ze3D00   1ze3     [1, 125]  [(1, 9), (19, 125)]   \n",
       "\n",
       "                                               sequence  \n",
       "6271  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original sequence:\n",
      "DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGDSEQGIVPCLTRAQLASMGLNTASVAGMNLLADDACVPLTTMVQDATAHLDVGQQRLNLTIPQAFMSNRAR\n",
      "Sequence with gaps:\n",
      "GVALGATRVIYPAGQKQEQLAVTNNDENSTYLIQSWVENADGVKDGRFIVTPPLFAMKGKKENTLRILDATNNQLPQDRESLFWMNVKAIPSMDKSKLTENTLQLAIISRIKLYYRPAKLALPPDQAAEKLRFRRSANSLTLINPTPYYLTVTELNAGTRVLENALVPPMGESTVKLPSDAGSNITYRTINDYGALTPKMTGVMETGGCDVSARDVTVTLPDYPGSVPIPLTVYCAKSQNLGYYLSGTTADAGNSIFTNTASFSPAQGVGVQLTRNGTIIPANNTVSLGAVGTSAVSLGLTANYARTGGQVTAGNVQSIIGVTFVYQDLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGDSEQGIVPCLTRAQLASMGLNTASVAGMNLLADDACVPLTTMVQDATAHLDVGQQRLNLTIPQAFMSNRAR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 3873.\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain H is discontinuous at line 3905.\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 3917.\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 3921.\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain H is discontinuous at line 4143.\n",
      "  warnings.warn(\n",
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 4276.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "random_pdb = '1ze3'\n",
    "\n",
    "print('The sequence with gaps info:')\n",
    "df = miss_residue_pdbs[miss_residue_pdbs['pdb_id'] == random_pdb][['cath_id', 'pdb_id', 'domain_edges', 'piece_edges', 'sequence']]\n",
    "display(df)\n",
    "\n",
    "print('Original sequence:')\n",
    "orig_seq = df['sequence'].values[0]\n",
    "print(orig_seq)\n",
    "\n",
    "print('Sequence with gaps:')\n",
    "new_seq = get_sequence_from_pdb_file(f\"{DATA_HOME}/pdb_downloaded/{random_pdb}.pdb\")\n",
    "print(new_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the second sequence is much longer.. So we need to align them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------D-----L---------------------------------------------------------------------------------Y------------------F----N------P--------------R------------------------------------------------------------------------------------------------F---------------L--------------L-----S-----------R-----------------F----------------ENGQELPPGTYRVDIYLNNGYMATRDVTFNTGDSEQGIVPCLTRAQLASMGLNTASVAGMNLLADDACVPLTTMVQDATAHLDVGQQRLNLTIPQAFMSNRAR\n",
      "                         |     |                                                                                 |                  |    |      |              |                                                                                                |               |              |     |           |                 |                |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GVALGATRVIYPAGQKQEQLAVTNNDENSTYLIQSWVENADGVKDGRFIVTPPLFAMKGKKENTLRILDATNNQLPQDRESLFWMNVKAIPSMDKSKLTENTLQLAIISRIKLYYRPAKLALPPDQAAEKLRFRRSANSLTLINPTPYYLTVTELNAGTRVLENALVPPMGESTVKLPSDAGSNITYRTINDYGALTPKMTGVMETGGCDVSARDVTVTLPDYPGSVPIPLTVYCAKSQNLGYYLSGTTADAGNSIFTNTASFSPAQGVGVQLTRNGTIIPANNTVSLGAVGTSAVSLGLTANYARTGGQVTAGNVQSIIGVTFVYQDLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGDSEQGIVPCLTRAQLASMGLNTASVAGMNLLADDACVPLTTMVQDATAHLDVGQQRLNLTIPQAFMSNRAR\n",
      "  Score=116\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/I0550745/miniconda3/envs/protein-pred/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "def align_sequences(s1, s2):\n",
    "    alignments = pairwise2.align.globalxx(s1, s2)\n",
    "    print(format_alignment(*alignments[0]))\n",
    "\n",
    "align_sequences(orig_seq, new_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the right part of the sequence perfectly aligns.\n",
    "\n",
    "*However, it seems like a lot of work to match new sequences with old ones so we will take the sequences as they are.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's create the target class, then save the data\n",
    "\n",
    "We will save `target` as merged `{class}.{architecture}`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Target distribution\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "target\n",
       "2.30    640\n",
       "3.90    634\n",
       "2.40    634\n",
       "3.40    628\n",
       "2.60    628\n",
       "1.10    626\n",
       "1.20    626\n",
       "3.30    626\n",
       "3.10    614\n",
       "3.20    612\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>sequence</th>\n",
       "      <th>piece_edges</th>\n",
       "      <th>num_residues</th>\n",
       "      <th>domain_edges</th>\n",
       "      <th>total_num_residues</th>\n",
       "      <th>num_gaps</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>123</td>\n",
       "      <td>[2, 124]</td>\n",
       "      <td>123</td>\n",
       "      <td>0</td>\n",
       "      <td>3.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>3be3</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 49), (51, 81)]</td>\n",
       "      <td>75</td>\n",
       "      <td>[6, 81]</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "      <td>2.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 492), (501, 555)]</td>\n",
       "      <td>99</td>\n",
       "      <td>[449, 555]</td>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "      <td>3.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1peq</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>53</td>\n",
       "      <td>[294, 346]</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>1.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>80</td>\n",
       "      <td>[77, 156]</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6263</th>\n",
       "      <td>2yyiA02</td>\n",
       "      <td>2yyi</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>10</td>\n",
       "      <td>1.66</td>\n",
       "      <td>ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...</td>\n",
       "      <td>[(139, 196), (199, 266)]</td>\n",
       "      <td>126</td>\n",
       "      <td>[139, 266]</td>\n",
       "      <td>128</td>\n",
       "      <td>1</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6264</th>\n",
       "      <td>4mo0A00</td>\n",
       "      <td>4mo0</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>780</td>\n",
       "      <td>10</td>\n",
       "      <td>2.10</td>\n",
       "      <td>EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...</td>\n",
       "      <td>[(24, 102)]</td>\n",
       "      <td>79</td>\n",
       "      <td>[24, 102]</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>3.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>1vq8X00</td>\n",
       "      <td>1vq8</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>440</td>\n",
       "      <td>10</td>\n",
       "      <td>2.20</td>\n",
       "      <td>ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...</td>\n",
       "      <td>[(7, 88)]</td>\n",
       "      <td>82</td>\n",
       "      <td>[7, 88]</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>3.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6266</th>\n",
       "      <td>1ze3D00</td>\n",
       "      <td>1ze3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>1.84</td>\n",
       "      <td>DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...</td>\n",
       "      <td>[(1, 9), (19, 125)]</td>\n",
       "      <td>116</td>\n",
       "      <td>[1, 125]</td>\n",
       "      <td>125</td>\n",
       "      <td>1</td>\n",
       "      <td>3.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6267</th>\n",
       "      <td>3ejvA00</td>\n",
       "      <td>3ejv</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>450</td>\n",
       "      <td>50</td>\n",
       "      <td>1.40</td>\n",
       "      <td>TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...</td>\n",
       "      <td>[(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...</td>\n",
       "      <td>152</td>\n",
       "      <td>[2, 160]</td>\n",
       "      <td>159</td>\n",
       "      <td>6</td>\n",
       "      <td>3.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6268 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id pdb_id  class  architecture  topology  superfamily  \\\n",
       "0     2w3sB01   2w3s      3            90      1170           50   \n",
       "1     3be3A00   3be3      2            30        30          320   \n",
       "2     3zq4C03   3zq4      3            10        20          580   \n",
       "3     1peqA03   1peq      1            10      1650           20   \n",
       "4     1bdoA00   1bdo      2            40        50          100   \n",
       "...       ...    ...    ...           ...       ...          ...   \n",
       "6263  2yyiA02   2yyi      2            40       110           10   \n",
       "6264  4mo0A00   4mo0      3            30       780           10   \n",
       "6265  1vq8X00   1vq8      3            10       440           10   \n",
       "6266  1ze3D00   1ze3      3            10        20          410   \n",
       "6267  3ejvA00   3ejv      3            10       450           50   \n",
       "\n",
       "      resolution_in_angstroms  \\\n",
       "0                        2.60   \n",
       "1                        2.04   \n",
       "2                        3.00   \n",
       "3                        2.80   \n",
       "4                        1.80   \n",
       "...                       ...   \n",
       "6263                     1.66   \n",
       "6264                     2.10   \n",
       "6265                     2.20   \n",
       "6266                     1.84   \n",
       "6267                     1.40   \n",
       "\n",
       "                                               sequence  \\\n",
       "0     SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1     QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2     DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3     DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4     EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "...                                                 ...   \n",
       "6263  ATTHALTNPQVNRARPPSGQPDPYIPVGVVKQTEKGIVVRGARMTA...   \n",
       "6264  EQKIKIYVTKRRFGKLMTIIEGFDTSVIDLKELAKKLKDICACGGT...   \n",
       "6265  ERVVTIPLRDARAEPNHKRADKAMILIREHLAKHFSVDEDAVRLDP...   \n",
       "6266  DLYFNPRFLLSRFENGQELPPGTYRVDIYLNNGYMATRDVTFNTGD...   \n",
       "6267  TADETIILNVLGQYTRAHDRRDPDAAALFAPEATIEIVDAVGGASR...   \n",
       "\n",
       "                                            piece_edges  num_residues  \\\n",
       "0                                            [(2, 124)]           123   \n",
       "1                                   [(6, 49), (51, 81)]            75   \n",
       "2                              [(449, 492), (501, 555)]            99   \n",
       "3                                          [(294, 346)]            53   \n",
       "4                                           [(77, 156)]            80   \n",
       "...                                                 ...           ...   \n",
       "6263                           [(139, 196), (199, 266)]           126   \n",
       "6264                                        [(24, 102)]            79   \n",
       "6265                                          [(7, 88)]            82   \n",
       "6266                                [(1, 9), (19, 125)]           116   \n",
       "6267  [(2, 2), (4, 27), (29, 66), (69, 99), (101, 14...           152   \n",
       "\n",
       "     domain_edges  total_num_residues  num_gaps target  \n",
       "0        [2, 124]                 123         0   3.90  \n",
       "1         [6, 81]                  76         1   2.30  \n",
       "2      [449, 555]                 107         1   3.10  \n",
       "3      [294, 346]                  53         0   1.10  \n",
       "4       [77, 156]                  80         0   2.40  \n",
       "...           ...                 ...       ...    ...  \n",
       "6263   [139, 266]                 128         1   2.40  \n",
       "6264    [24, 102]                  79         0   3.30  \n",
       "6265      [7, 88]                  82         0   3.10  \n",
       "6266     [1, 125]                 125         1   3.10  \n",
       "6267     [2, 160]                 159         6   3.10  \n",
       "\n",
       "[6268 rows x 14 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_data = proc_data.assign(\n",
    "    target=lambda df: df[\"class\"].astype(str) + \".\" + df[\"architecture\"].astype(str)\n",
    ").reset_index(drop=True)\n",
    "\n",
    "print('Target distribution')\n",
    "display(out_data['target'].value_counts())\n",
    "\n",
    "out_data.to_csv(f\"{DATA_HOME}/data_processed.csv\")\n",
    "out_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "protein-pred",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
